### ISSUE TITLE
CMake doesn't properly generate when LTO is used

### ISSUE THREAD

**Issue by np511** [PCSX2/pcsx2#1671]
***PCSX2 version:*** development build 1.5.0-dev-1638-g0d27586

***PCSX2 options:***
N/A

***Plugins used:***
N/A

***Description of the issue:***
When the USE_LTO option is selected, compilation fails. After doing some testing, I found that the problem is actually multiple problems:

* For some odd reason, the generated link.txt files have "--plugin /usr/lib/gcc/x86_64-linux-gnu/4.9/" and then something along the lines of "lto_plugin.so". I manually fixed the problem so I can't say for certain. This might be acceptable - if GCC 4.9 was installed on my computer (it's not). I had to carefully run an find -exec sed type command to fix the problem without causing CMake to regenerate the files.
* Even when I specify gcc-ar or /usr/bin/gcc-ar under CMAKE-AR, /usr/bin/ar is used. LTO does *not* work with the regular ar program. gcc-ar is needed.
* As with ar, ranlib had to be manually adjusted to gcc-ranlib with sed and find. I did the same for nm just in case it was used somewhere since I do not actually  know what nm does.
* The SPU2x, ZZogl, and LilyPad (and possibly others) fail to compile when both LTO and the "gold" linker are used. I am not entirely certain why, but there was some message about a relocation. Oddly, removing "-fuse-ld=gold" for these plugins (both under link.txt and flags.make) resulted in successful compilation. I did not try removing gold only from the link.txt. I did not try removing gold only from the flags.make. 
* Additionally, -flto=4 is used by default. On my computer, -flto=2 would probably run faster. If there was a way to specify this in CMake, it would make the process easier as well.

***How to reproduce the issue:***

***Last known version to work:***
I was unable to use LTO either in stable or development at any time without manual modification of the files generated by CMake.

***PC specifications:***
CPU: Intel Celeron N3050 @1.60 GHz  Dual-Core  OS: Debian Linux 64-bit (It's not caused by multi-arch issues, it compiles fine with the supplied toolchain file when LTO is disabled).

Going overboard here, but the output of "cat /proc/cpuinfo":

processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 76
model name      : Intel(R) Celeron(R) CPU  N3050  @ 1.60GHz
stepping        : 3
microcode       : 0x35e
cpu MHz         : 479.980
cache size      : 1024 KB
physical id     : 0
siblings        : 2
core id         : 0
cpu cores       : 2
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 11
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 cx16 xtpr pdcm sse4_1 sse4_2 movbe popcnt tsc_deadline_timer aes rdrand lahf_lm 3dnowprefetch epb tpr_shadow vnmi flexpriority ept vpid tsc_adjust smep erms dtherm ida arat
bugs            :
bogomips        : 3200.00
clflush size    : 64
cache_alignment : 64
address sizes   : 36 bits physical, 48 bits virtual
power management:

processor       : 1
vendor_id       : GenuineIntel
cpu family      : 6
model           : 76
model name      : Intel(R) Celeron(R) CPU  N3050  @ 1.60GHz
stepping        : 3
microcode       : 0x35e
cpu MHz         : 479.980
cache size      : 1024 KB
physical id     : 0
siblings        : 2
core id         : 2
cpu cores       : 2
apicid          : 4
initial apicid  : 4
fpu             : yes
fpu_exception   : yes
cpuid level     : 11
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 cx16 xtpr pdcm sse4_1 sse4_2 movbe popcnt tsc_deadline_timer aes rdrand lahf_lm 3dnowprefetch epb tpr_shadow vnmi flexpriority ept vpid tsc_adjust smep erms dtherm ida arat
bugs            :
bogomips        : 3202.57
clflush size    : 64
cache_alignment : 64
address sizes   : 36 bits physical, 48 bits virtual
power management:

**Comment by gregory38** [comment]
Don't use LTO, my gcc is limited to 4.9 (hence the hardcoded path) and I didn't manage to make it work.

**Comment by np511** [comment]
On Tuesday, November 22, 2016 7:48:20 AM EST Gregory Hainaut wrote:
> Don't use LTO, my gcc is limited to 4.9 (hence the hardcoded path) and I
> didn't manage to make it work.
I got LTO to work - I just had to manually edit some of the files CMake 
generates. The problem is that every time the source is updated, I'll have to 
run a shell script to fix the generated files. CMake should properly work.

Regardless of what version of gcc you have, the "--plugin" shouldn't be added. 
When LTO is used "gcc-ar" should be used instead of "ar", "gcc-ranlib" instead 
of "ranlib", and "gcc-nm" instead of "nm". If those changes are added, the 
majority of PCSX2 compiles perfectly.

Remove -fuse-ld=gold from a few of the plugins and there is no problem.

**Comment by gregory38** [comment]
You might need to edit those lines
https://github.com/PCSX2/pcsx2/blob/master/cmake/BuildParameters.cmake#L334

Feel free to post any working update.

**Comment by np511** [comment]
On Tuesday, November 22, 2016 8:06:07 AM EST Gregory Hainaut wrote:
> You might need to edit those lines
> https://github.com/PCSX2/pcsx2/blob/master/cmake/BuildParameters.cmake#L334
> 
> Feel free to post any working update.
Thank you! I've been meaning to learn CMake for a while now, but haven't 
gotten around to it yet. This is exactly what I needed!

**Comment by np511** [comment]
On Tuesday, November 22, 2016 8:06:07 AM EST Gregory Hainaut wrote:
> You might need to edit those lines
> https://github.com/PCSX2/pcsx2/blob/master/cmake/BuildParameters.cmake#L334
> 
> Feel free to post any working update.

I've never used git for any project other than personal ones. How do I submit 
a working patch? I am familiar with git to the point of knowing most of the 
commands and what they do, but not how to submit to someone else's project.

I can just upload a .patch file if so desired.

Either way, I'm testing the patch now. I created a patch file anyway so I don't 
need to worry about starting over if it goes south.

**Comment by np511** [comment]
On Tuesday, November 22, 2016 8:06:07 AM EST Gregory Hainaut wrote:
> You might need to edit those lines
> https://github.com/PCSX2/pcsx2/blob/master/cmake/BuildParameters.cmake#L334
> 
> Feel free to post any working update.
Just a thought, but when using multilib, would it be possible to create x86_64 
plugins, a x86_64 plugin loader which communicates with the ix86 executable, 
and the main executable in ix86? Even if the main emulator can't be easily 
ported to 64-bit, some of the plugins probably can be.

Alternatively, would it be possible to make PCSX2 64-bit and use a small 32-
bit executable to actually run the generated code? Compiling with multilib is 
very, very annoying. Especially with dependency conflicts on Debian that 
prevent me from having libsdl2-dev as 64-bit and 32-bit at once.

**Comment by mirh** [comment]
I'm pretty sure gregory already know (and feels) this concerns. 
The plan should be to have anything compile with either instruction set one day and he's indeed working on the gsdx port (I can't comment on the state of patches to port core too though

> I've never used git for any project other than personal ones. How do I submit a working patch?

Do whatever is needed to a new branch in a personal repository, then from github website open a pull request.

**Comment by np511** [comment]
On Tuesday, November 22, 2016 11:47:41 AM EST mirh wrote:
> I'm pretty sure gregory already know (and feels) this concerns.
> The plan should be to have anything compile with either instruction set one
> day and he's indeed working on the gsdx port (I can't comment on the state
> of patches to port core too though
> > I've never used git for any project other than personal ones. How do I
> > submit a working patch?
> Do whatever is needed to a new branch in a personal repository, then from
> github website open a pull request.
Thanks! I suspected as much, but I felt it was better safe than sorry.

I'll just wait a while to make sure compilation succeeds, then I'll get to 
work on that. Currently, I think I've got CMake to generate the files properly.

Unfortunately, I had to remove the "-fuse-ld=gold" line as a few plugins fail 
(as I mentioned in the OP). However, what seems strange to me is that gold was 
used for LTO but not without it. Was there a reason for this?

I'll see if I can get CMake to use gold only for the main executable and the 
plugins that support it, but I probably won't put as much effort into it as I 
would if I knew why gold was needed.

It seems to be succeeding so far, but I'm waiting for completion to start work 
on GitHub.

**Comment by gregory38** [comment]
```
Just a thought, but when using multilib, would it be possible to create x86_64 plugins, a x86_64 plugin loader which communicates with the ix86 executable, and the main executable in ix86? Even if the main emulator can't be easily ported to 64-bit, some of the plugins probably can be. Alternatively, would it be possible to make PCSX2 64-bit and use a small 32- bit executable to actually run the generated code? Compiling with multilib is very, very annoying. Especially with dependency conflicts on Debian that prevent me from having libsdl2-dev as 64-bit and 32-bit at once.
```
Frankenstein solution isn't a good idea and it is quite complex.

Let's say you keep the recompiler in 32 bits, how do you access C object with a 64 bits address but 32 bits register. So all the core must remain 32 bits. Now if you want 64 bit plugins, you can't directly link them the ABI isn't compatible. Let's imagine that we manage to plug various ASM to make ABI compatible. You might need to reprogram your CPU. Because the CPU run either in 32 bits or in 64 bits mode. You can't switch it every function call.

In short, it is all or nothing. 

```
Compiling with multilib is very, very annoying.
```
I know. I recompiled all GTK2 stack + mesa + qapitrace + various dependency on my system. I think you can build without SDL, you will miss onepad plugin but you can still keep the debian version.

**Comment by gregory38** [comment]
Shall we close the bug ?

**Comment by np511** [comment]
On Tuesday, November 29, 2016 2:36:43 AM EST Gregory Hainaut wrote:
> Shall we close the bug ?
I don't see why not. I've got PGO working, and will make a PR later. I don't 
have time now.

**Comment by mirh** [comment]
*#1678*
